lib/helpers.go

```diff
@@ -0,0 +1,153 @@
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package lib
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+)
+
+// StrictJSONUnmarshal decodes a JSON in a strict manner, emitting an error if there
+// are unknown fields or unexpected data
+func StrictJSONUnmarshal(data []byte, v interface{}) error {
+	dec := json.NewDecoder(bytes.NewReader(data))
+	dec.DisallowUnknownFields()
+	dec.UseNumber()
+
+	if err := dec.Decode(&v); err != nil {
+		return err
+	}
+	if dec.More() {
+		//TODO: use a custom error?
+		return fmt.Errorf("unexpected data after the JSON object")
+	}
+	return nil
+}
+
+// GetMaxPlannedVUs returns the maximum number of planned VUs at any stage of
+// the scheduler execution plan.
+func GetMaxPlannedVUs(steps []ExecutionStep) (result uint64) {
+	for _, s := range steps {
+		stepMaxPlannedVUs := s.PlannedVUs
+		if stepMaxPlannedVUs > result {
+			result = stepMaxPlannedVUs
+		}
+	}
+	return result
+}
+
+// GetMaxPossibleVUs returns the maximum number of planned + unplanned (i.e.
+// initialized mid-test) VUs at any stage of the scheduler execution plan.
+// Unplanned VUs are possible in some schedulers, like the arrival-rate ones, as
+// a way to have a low number of pre-allocated VUs, but be able to initialize
+// new ones in the middle of the test, if needed. For example, if the remote
+// system starts responding very slowly and all of the pre-allocated VUs are
+// waiting for it.
+//
+// IMPORTANT 1: Getting planned and unplanned VUs separately for the whole
+// duration of a test can often lead to mistakes. That's why this function is
+// called GetMaxPossibleVUs() and why there is no GetMaxUnplannedVUs() function.
+//
+// As an example, imagine that you have a scheduler with MaxPlannedVUs=20 and
+// MaxUnaplannedVUs=0, followed immediately after by another scheduler with
+// MaxPlannedVUs=10 and MaxUnaplannedVUs=10. If you MaxPlannedVUs for the whole
+// test is 20, and MaxUnaplannedVUs, but since those schedulers won't run
+// concurrently, MaxVUs for the whole test is not 30, rather it's 20, since 20
+// VUs will be sufficient to run the test.
+//
+// IMPORTANT 2: this has one very important exception. The manual execution
+// scheduler doesn't use the MaxUnplannedVUs (i.e. this function will return 0),
+// since their initialization and usage is directly controlled by the user and
+// is effectively bounded only by the resources of the machine k6 is running on.
+func GetMaxPossibleVUs(steps []ExecutionStep) (result uint64) {
+	for _, s := range steps {
+		stepMaxPossibleVUs := s.PlannedVUs + s.MaxUnplannedVUs
+		if stepMaxPossibleVUs > result {
+			result = stepMaxPossibleVUs
+		}
+	}
+	return result
+}
+
+// GetEndOffset returns the time offset of the last step of the execution plan,
+// and whether that step is a final one, i.e. whether the number of planned or
+// unplanned
```

Comment:
I don't understand the part after `i.e.` I suppose there was suppose to be more text after `unplanned` ?

Reply:
yep, thanks, will fix!